### DISCUSSION TITLE
Is there a way to detect and respond to changes in validation errors or error count other than relying on the "errors" object, which seems to sneakily mutate?

### DISCUSSION THREAD

**Discussion by alexkatz** [react-hook-form/react-hook-form#4905] [Q&A]
It appears as though react-hook-form's `errors` object is mutated rather than replaced when validation errors are updated. This makes it a bit difficult to detect when validation errors have changed, especially in a large form or a field array form of dynamic length.

For example, let's say I want to log all validation errors whenever there is a change in error count. The type of approach below does not work, since the `errors` object is mutated by react-hook-form rather than replaced when validation error state changes:

```
useEffect(() => {
// as expected, this does not work, and only logs once
  console.log('errors has changed:', errors);
}, [errors]);
```

Currently the only way I can see to do this is to stringify the errors object to detect changes, as so:

```
const errorsStringified = JSON.stringify(errors);
useEffect(() => {
  // this sucks, but it works
  console.log('errorsStringified has changed:', errors);
}, [errors, errorsStringified]);
```

Let's say I need to display an accurate custom error summary of some kind. Is there a way to have a component track errors as they grow or shrink that doesn't require manually parsing a constantly mutating `errors` object? 

It seems to me like react-hook-form could return a newly instantiated `errors` object when one or more errors change, or return the same object from the previous render if no validation changes are detected. This would allow users to reliably use `errors` as a hook dependency and doesn't seem like it would change much else. Is there a reason the library does not already work this way? 

Thanks!

**Comment by bluebill1049** [comment]
you should put `formState` as `dep`, we don't memory individual form states. Early in the morning, i hope that's the answer your are seeking for.

**Reply by alexkatz** [reply]
So if one needed to display an up-to-date error count of a form, for example, there's no way to know if the number of errors has changed other than to parse the entirety of the `errors` object on every single render? 

I'm struggling to find another way, but parsing or stringifying `errors` seems to be the only possible way given the instance of the `errors` object seems to be the same across renders and is simply mutated.

By the way, `react-hook-form` is great! I can't thank you enough for such an awesome library.

**Reply by bluebill1049** [reply]
thank you for your kind words @alexkatz, yes you can, you just need to use a deepCompare hook, I think kent have built one. give it a try.

**Reply by bluebill1049** [reply]
```
useDeepCompare(() => {}, [errors])
```

something like above
